### PR TITLE
add AutoMixedPrecision graph optimization pass

### PR DESCRIPTION
**Pull request by minminsun**
This pull request is to add an AutoMixedPrecision graph optimization pass, which automatically transforms a normal tensorflow graph to mixed-precision graph, in order to take advantage of hardware accelerators like TensorCore of Volta GPU.
Although NVIDIA provided documents and sample codes on how to write tensorflow models in mixed-precision (https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html#tensorflow), but it's not convenient for most users, especially that most existing tensorflow models are written in fp32 and the effort to manually convert them to mixed precision models is too high.
With AutoMixedPrecision graph optimization pass, running mixed-precision training is as easy as adding a single configuration like "session_config.graph_options.optimizer_options.auto_mixed_precision=True" (In some cases, loss scaling e.g. tf.contrib.mixed_precision.LossScaleOptimizer is needed to avoid underflow of fp16 gradients).
The AutoMixedPrecision graph optimization pass has been widely used by various production groups on Alibaba-PAI (Platform of AI) for half years.

### CODE REVIEW

**Comment by yangjunpro** [comment]
This AutoMixedPrecision feature has already been successfully deployed onto Alibaba PAI platform running for more than half a year  and lots of our internal workload benefits from this feature since it significantly release the laboring work for manually converting fp32 models to fp16 version, which is both time-consuming and error-prone.  
This feature has been applied for diversified workloads, including image recognition, image detection, machine translate, BERT, language model, large-scale embedding model and graph embedding models. The performance speed-up ranges from 1.3X to 3X with the same convergence trend which truly help exploit the computing power of NV's TensorCore accelerator.

Also there is a related PR regarding to [this one](https://github.com/tensorflow/tensorflow/pull/23244), inside Alibaba, this PR actually work together with our loss scaling manager to ensure our users have transparent experience using TensorCore feature.  Due to that the loss scaling manager PR may involves changes with the core gradient logic, several months ago we have discussed with the TF community and there are still some different design considerations, so [this PR](https://github.com/tensorflow/tensorflow/pull/23244)) is still pending.

Since the AutoMixedPrecision graph optimization pass can be decoupled from the automatic loss scaling manager feature, we make this separate PR wishing community could benefit from our internal work earlier. The AutoMixedPrecision graph optimization pass within this PR actually could cooperate with the tf.contrib.mixed_precision.LossScaleOptimizer to ensure both performance speed-up as well as convergence guarantee. Although personally we think what is implemented in [this PR](https://github.com/tensorflow/tensorflow/pull/23244) might be a better choice than tf.contrib.mixed_precision.LossScaleOptimizer since it make users' life easier:).

Any feedbacks from community are highly welcome.

**Comment by yangjunpro** [comment]
@rthadur 

Thanks for the prompt follow-up, any comments and feedback are highly welcome.

**Comment by gunan** [comment]
I am not the right reviewer for this PR, please redirect.

**Comment by yangjunpro** [comment]
@rthadur @mrry @reedwm 
Folks,

Any comments or feedback from your side?

Thanks

Jun

**Comment by yangjunpro** [comment]
> @rthadur @mrry @reedwm
> Folks,
> 
> Any comments or feedback from your side?
> 
> Thanks
> 
> Jun

@rthadur @mrry  @reedwm 
Hi, any update from your side? This PR has been pending for more than one weeks and we would like to know any kinds of feedback from TF community.

Regards.

**Comment by mrry** [comment]
I'm definitely not the right reviewer for this PR.

CC'ing @rmlarsen ... even though this isn't Grappler-related, he might know of a better reviewer.

**Comment by reedwm** [comment]
Sorry for the delayed response. @nluehr from Nvidia has also been working on a Grappler pass that enables mixed precision by making certain nodes fp16. We do not want two Grappler passes that have the same functionality. I hear you are planning on talking with @nluehr soon, and so if he can make his PR meet all your requirements, we will submit only his PR.

Feel free email me at <my username> AT google if you want to talk to me privately.

**Comment by yangjunpro** [comment]
> Sorry for the delayed response. @nluehr from Nvidia has also been working on a Grappler pass that enables mixed precision by making certain nodes fp16. We do not want two Grappler passes that have the same functionality. I hear you are planning on talking with @nluehr soon, and so if he can make his PR meet all your requirements, we will submit only his PR.
> 
> Feel free email me at AT google if you want to talk to me privately.

We will have communication with NV folks tomorrow about the collaboration. And personally I think it should be clarified about "only submit NV's PR" since our work certainly started much earlier than NV folks. It should be trivial for us to port our code from graph optimization to grappler pass, as what we have done for [the LINM grapper pass](https://github.com/tensorflow/tensorflow/commit/84fe908258550e1ce27e8725de1e2af279479c9d).  For the details I will communicate with you privately.

**Comment by nluehr** [comment]
Nvidia and Alibaba are collaborating to combine our graph pass algorithms into a single PR, for the initial version, which we expect will be ready within a few weeks.

**Comment by rthadur** [comment]
@rmlarsen @meteorcloudy can you please help review this PR

**Comment by reedwm** [comment]
I believe Alibaba has collaborated with Nvidia to create #26342, which contains the functionality of this PR. So I'll close this PR.

**Comment by venuswu** [comment]
How can I get the graph for tensorboard after mixed precision? @yangjunpro @minminsun

**Comment by reedwm** [comment]
Looks like you figured it out in tensorflow/models#7480. We can discuss the problem in that issue.
